server/gameserver.js

```diff
@@ -88,6 +88,82 @@ async function SharedTryAddServer( request )
 	}
 }
 
+async function TryReviveServer( request )
+{
+	let ghost = GetGhostServer( request.query.id )
+
+	if( request.ip != ghost.ip ) return
+
+	// check server's verify endpoint on their auth server, make sure it's fine
+	// in the future we could probably check the server's connect port too, with a c2s_connect packet or smth, but atm this is good enough
+	let modInfo
+
+	if ( request.isMultipart() )
+	{
+		try
+		{
+			modInfo = JSON.parse( ( await ( await request.file() ).toBuffer() ).toString() )
+		}
+		catch ( ex )
+		{
+			return { success: false, error: JSON_PARSE_ERROR }
+		}
+	}
+
+	let authServerResponse = await asyncHttp.request( {
+		method: "GET",
+		host: request.ip,
+		port: ghost.authPort,
+		path: "/verify"
+	} )
+
+	if ( !authServerResponse || authServerResponse.toString() != VERIFY_STRING )
+		return { success: false, error: NO_GAMESERVER_RESPONSE }
+
+	// pdiff stuff
+	if ( modInfo && modInfo.Mods )
+	{
+		for ( let mod of modInfo.Mods )
+		{
+			if ( mod.pdiff )
+			{
+				try
+				{
+					let pdiffHash = crypto.createHash( "sha1" ).update( mod.pdiff ).digest( "hex" )
+					mod.pdiff = pjson.ParseDefinitionDiffs( mod.pdiff )
+					mod.pdiff.hash = pdiffHash
+				}
+				catch ( ex )
+				{
+					mod.pdiff = null
+				}
+			}
+		}
+	}
+
+	let playerCount = request.query.playerCount || 0
+	if ( typeof playerCount == "string" )
+		playerCount = parseInt( playerCount )
+
+	if ( typeof request.query.maxPlayers == "string" )
+		request.query.maxPlayers = parseInt( request.query.maxPlayers )
+
+	let name = filter.clean( request.query.name )
+	let description = request.query.description == "" ? "" : filter.clean( request.query.description )
+	let newServer = new GameServer( name, description, playerCount, request.query.maxPlayers, request.query.map, request.query.playlist, request.ip, ghost.port, ghost.authPort, request.query.password, modInfo )
```

Comment:
You should refactor this section into a common function since this is basically a copypaste from `SharedTryAddServer`